src/Controller/Traits/PasswordManagementTrait.php

```diff
@@ -31,29 +31,41 @@ trait PasswordManagementTrait
 
     /**
      * Change password
+     * Can be used while logged in for own password, as a superuser on any user, or while not logged in for reset
+     * reset password with session key (email token has already been validated)
+     *
+     * @param int|string|null $id user_id, null for logged in user id
      *
      * @return mixed
      */
-    public function changePassword()
+    public function changePassword($id = null)
     {
         $user = $this->getUsersTable()->newEntity();
-        $id = $this->Auth->user('id');
-        if (!empty($id)) {
-            $user->id = $this->Auth->user('id');
-            $validatePassword = true;
-            //@todo add to the documentation: list of routes used
-            $redirect = Configure::read('Users.Profile.route');
+        if ($this->Auth->user('id')) {
+            if ($id && $this->Auth->user('is_superuser') && Configure::read('Users.Superuser.allowedToChangePasswords')) {
+                // superuser editing any account's password
+                $user->id = $id;
+                $validatePassword = false;
+                $redirect = ['action' => 'index'];
+            } elseif (!$id || $id === $this->Auth->user('id')) {
+                // normal user editing own password
+                $user->id = $this->Auth->user('id');
+                $validatePassword = true;
+                $redirect = Configure::read('Users.Profile.route');
+            } else {
+                throw new NotFoundException('Changing another user\'s password is not allowed');
```

Comment:
Instead of throwing an exception we should redirect the user with a flash message